### DISCUSSION TITLE
Reset Form Not Working w/ Controlled Components

### DISCUSSION THREAD

**Discussion by shoxter** [react-hook-form/react-hook-form#5464] [Q&A] [answered]
**Describe the bug**
I have a simple `FormTextField` (code below) that uses the control passed in from its parent from `useForm`. When resetting the form with `keepErrors` and `keepValid` both set to `false`, the errors remain in both the `formState.errors` and the controller's `fieldState.error` properties. Additionally, calling `clearErrors` after the `reset` doesn't fix the issue either.

FormTextFieldCode:

```
function FormTextField(props) {
	const {formProps, ...restProps} = props;
	const {field, fieldState} = useController({
		name: formProps.name,
		control: formProps.control
	});

	return (
		<TextField
			error={!!fieldState.error}
			helperText={fieldState.error?.message}
			{...restProps}
			inputRef={field.ref}
			onChange={(e) => {
				field.onChange(e.target.value);
				if (restProps.onChange) {
					restProps.onChange(e);
				}
			}}
			onBlur={(e) => {
				field.onBlur(e.target.value);
				if (restProps.onBlur) {
					restProps.onBlur(e);
				}
			}}
			value={field.value}
		/>
	);
}
```

Reset code not removing errors:

```
	reset(null, {
		keepErrors: false,
		keepDirty: false,
		keepDefaultValues: true,
		keepIsSubmitted: false,
		keepTouched: false,
		keepIsValid: false,
		keepSubmitCount: false
	});
```

**To Reproduce**

1. Create FormTextField above.
2. Pass control and name from parent component.
3. Enter something invalid.
3. Call reset code.
4. See error remain.

**Codesandbox link (Required)**

*Note* -- The isDirty property here is wrong too initially causing the onBlur to not function properly. I'm not sure why after blurring it still thinks the form isn't dirty. To replicate the error issue I mention above:

1. Set the value to -3 (blur)
2. Set the value to 1 (blur)
3. Set the value back to -3 (blur)
4. The value will reset to a valid value but errors remain even though reset is saying not to keep them.

https://codesandbox.io/s/dank-dew-gdt5m

**Expected behavior**
Upon calling `reset` with the above values, the expected behavior would be for the form to not have errors. It almost seems like there is a lot of async code behind the scenes fighting to update these properties and when they're not executing in the right order.

**Screenshots**

**Desktop (please complete the following information):**

- OS: Arch Linux
- Browser: Chrome 91

**Comment by bluebill1049** [comment, marked as answer]
I wouldn't do that reset inside the event handler, the event handle is async  by default, instead, the reset should perhaps go into `useEffect`

**Reply by shoxter** [reply]
Thanks for the input Bill. Definitely appreciate your advice!

**Reply by avalanche1** [reply]
@bluebill1049 How is asynchronous nature of `handleSubmit` prevents `reset` from working as expected?
And if that is that important it should be stated explicitly in the docs
https://react-hook-form.com/api/useform/handlesubmit
I'm also having this kind of problem with reset not clearing errors and have to dig it up to here 😭 
Also, it is very unintuitive to put `reset` into `useEffect`; my natural desire is to put it into `handleSubmit`. Maybe, you should revise your architecture to conform to devs' natural instincts?

**Reply by avalanche1** [reply]
@shoxter Did you manage to get it working? I'm forced to use `clearErrors` manually, instead of relying on `reset()`.

**Reply by shoxter** [reply]
@avalanche1 We actually ended up re-implementing the entire library from scratch with almost an identical API so that it drop-in replaced all our RHF use cases. There were too many validation issues that we couldn't easily solve in simple use cases so we just decided to re-implement it ourselves.
